src/curl_jolokia.c

```diff
@@ -0,0 +1,1045 @@
+			       yajl_len_t *SetThisLen) {
+	if (!((ptr > ShouldBeInHere->buffer) &&
+	      (ptr < ShouldBeInHere->buffer + ShouldBeInHere->used))) {
+		char *tptr = BufferHereIfNot->buffer + BufferHereIfNot->used;
+		cjo_append_buffer(BufferHereIfNot, ptr, len);
+		ptr = tptr;
+	}
+
+	*SetThisOne = ptr;
+	*SetThisLen = (yajl_len_t) len;
+}
+
+/*
+ * llist lookup methods
+ */
+static int cjo_cfg_compare_attribute(llentry_t *match, void *vdb) {
+	cjo_t *db = (cjo_t *)vdb;
+	cjo_attribute_t *pcfg = (cjo_attribute_t *)match->value;
+
+	if (pcfg->attribute_match_len != db->curr_attribute->json_name_len)
+		return -1;
+	else
+		return strncmp(pcfg->attribute_match, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+}
+static int cjo_cfg_compare_bean(llentry_t *match, void *vdb) {
+	cjo_t *db = (cjo_t *)vdb;
+	cjo_bean_t *pbeancfg = (cjo_bean_t *)match->value;
+
+	if (pbeancfg->mbean_match_len != db->json_key_len)
+		return -1;
+	return strncmp(pbeancfg->mbean_match, db->json_key, db->json_key_len);
+}
+
+/*
+ * Aligning data with Configs & Sending it
+ */
+
+static int cjo_get_type(cjo_attribute_t *attribute) {
+	if (attribute->type != NULL) {
+		return -EINVAL;
+	}
+
+	const data_set_t *ds = plugin_get_ds(attribute->type);
+	if (ds == NULL) {
+		static char type[DATA_MAX_NAME_LEN] = "!!!invalid!!!";
+
+		if (strcmp(type, attribute->type) != 0) {
+			ERROR("curl_jolokia plugin: Unable to look up DS type \"%s\".",
+			      attribute->type);
+			sstrncpy(type, attribute->type, sizeof(type));
+		}
+
+		return -1;
+	} else if (ds->ds_num > 1) {
+		static c_complain_t complaint = C_COMPLAIN_INIT_STATIC;
+
+		c_complain_once(
+			LOG_WARNING, &complaint,
+			"curl_jolokia plugin: The type \"%s\" has more than one data source. "
+			"This is currently not supported. I will return the type of the "
+			"first data source, but this will likely lead to problems later on.",
+			attribute->type);
+	}
+
+	return ds->ds[0].type;
+}
+
+static void cjo_submit(cjo_t *db) /* {{{ */
+{
+	if ((db->match_this_bean == NULL) || (db->attribute_pool_used == 0))
+		return; /* nothing to do yet. */
+
+	for (int i = 0; i < db->attribute_pool_used; i++) {
+		value_list_t vl = VALUE_LIST_INIT;
+
+		db->curr_attribute = &db->attributepool[i];
+		llentry_t *cfgptr = llist_search_custom(db->match_this_bean->attributes,
+					     cjo_cfg_compare_attribute, db);
+		if (cfgptr == NULL) {
+			char attribute[db->curr_attribute->json_name_len + 1];
+
+			memcpy(attribute, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+			attribute[db->curr_attribute->json_name_len] = '\0';
+			ERROR("curl_jolokia plugin: failed to locate attribute [%s:\"%s\"]",
+			      db->match_this_bean->bean_name, attribute);
+
+			continue; /* we don't know this! */
+		}
+		cjo_attribute_t *curr_attribute = cfgptr->value;
+
+		/* Create a null-terminated version of the string. */
+		int ds_type = cjo_get_type(curr_attribute);
+		if (ds_type == -1) {
```

Comment:
`cjo_get_type()` may contain `-EINVAL` too, so `(ds_type < 0)` would be a better check.